operator/metrics/metrics.go

```diff
@@ -15,28 +15,119 @@
 package metrics
 
 import (
-	"log"
+	"context"
 	"net/http"
 
 	operatorOption "github.com/cilium/cilium/operator/option"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
 
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "metrics")
+)
+
 const Namespace = "cilium_operator"
 
 var (
-	Registry *prometheus.Registry
+	Registry   *prometheus.Registry
+	shutdownCh chan struct{}
 )
 
 func Register() {
+	log.Info("Registering Operator metrics")
+
 	Registry = prometheus.NewPedanticRegistry()
-	Registry.MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{Namespace: Namespace}))
+	registerMetrics()
+
+	m := http.NewServeMux()
+	m.Handle("/metrics", promhttp.HandlerFor(Registry, promhttp.HandlerOpts{}))
+	srv := &http.Server{
+		Addr:    operatorOption.Config.OperatorPrometheusServeAddr,
+		Handler: m,
+	}
+
+	shutdownCh = make(chan struct{})
 	go func() {
-		// The Handler function provides a default handler to expose metrics
-		// via an HTTP server. "/metrics" is the usual endpoint for that.
-		http.Handle("/metrics", promhttp.HandlerFor(Registry, promhttp.HandlerOpts{}))
-		log.Fatal(http.ListenAndServe(operatorOption.Config.OperatorPrometheusServeAddr, nil))
+		go func() {
+			err := srv.ListenAndServe()
+			switch err {
+			case http.ErrServerClosed:
+				log.Info("Metrics server shutdown successfully")
+				return
+			default:
+				log.Fatalf("Metrics server ListenAndServe failed: %v", err)
+			}
+		}()
+
+		<-shutdownCh
+		log.Info("Received shutdown signal")
+		if err := srv.Shutdown(context.TODO()); err != nil {
+			log.Errorf("Shutdown operator metrics server failed: %v", err)
+		}
 	}()
 }
+
+func Unregister() {
+	log.Info("Shutting down metrics server")
+
+	if shutdownCh == nil {
+		log.Error("Unregister metrics: shutdownCh == nil")
+		return
+	}
```

Comment:
Is logging this error here necessary? The `shutdownCh` is initialized when we call Register(), so if someone calls Unregister() before calling Register(), then there's nothing to shut down. I wouldn't consider that an error. We can simply just return here.

```suggestion
	if shutdownCh == nil {
		return
	}
```